ocaml/typing/typecore.ml

```diff
@@ -6953,52 +6983,38 @@ and type_cases
-                  match pc_guard with
-                  (* This case is unreachable, as [pc_guard] cannot match the
-                     outer pattern while also matching [Some (Guard_pattern _)]
+                  match pc_rhs with
+                  (* This case is unreachable, as [pc_rhs] cannot match the
+                     outer pattern while also matching [Ppattern_guarded_rhs _]
                    *)
-                  | Some (Guard_pattern _) -> assert false
-                  | None -> None
-                  | Some (Guard_predicate pred) ->
+                  | Ppattern_guarded_rhs _ -> assert false
+                  | Psimple_rhs _ -> None
+                  | Pboolean_guarded_rhs { pbg_guard; _ } ->
                       let expected_bool =
                         mk_expected ~explanation:When_guard Predef.type_bool
                       in
-                      Some
-                        (Predicate
-                          (type_expect ext_env mode_local pred expected_bool))
+                      let typed_guard =
+                        type_expect ext_env mode_local pbg_guard expected_bool
+                      in
+                      Some (Predicate typed_guard)
                 in
                 let exp =
-                  type_expect
-                    ?in_function ext_env emode pc_rhs
+                  type_expect ?in_function ext_env emode rhs
                     (mk_expected ?explanation ty_res')
                 in
                 guard, exp
-            | Some
-                (Guard_pattern 
-                  { pgp_scrutinee = e; pgp_pattern = pat; pgp_loc = loc }) ->
-                let { arg; sort; cases; partial; } =
+            | Ppattern_guarded_rhs
+                  { ppg_scrutinee; ppg_cases; ppg_loc = loc } ->
+                let { arg = _; sort = _; cases = _; partial = _; } =
                   type_match
-                    e [ { pc_lhs = pat; pc_guard = None; pc_rhs } ] ext_env loc
-                    Check_and_warn_if_total (mk_expected ?explanation ty_res')
-                    emode
+                    ppg_scrutinee ppg_cases ext_env loc Check_and_warn_if_total
+                    (mk_expected ?explanation ty_res') emode
                 in
-                (match cases with
-                  | [ { c_lhs = pat; c_guard = None; c_rhs = exp } ] ->
-                      let pattern_guard : Typedtree.guard =
-                        Pattern
-                           { pg_scrutinee = arg
-                           ; pg_scrutinee_sort = sort
-                           ; pg_pattern = pat
-                           ; pg_partial = partial
-                           ; pg_loc = loc
-                           ; pg_env = ext_env }
-                      in
-                      Some pattern_guard, exp
-                  | _ ->
-                      Misc.fatal_error "type_cases invariant violated")
+                fatal_error "typechecking for multicase pattern guards \
```

Comment:
I think this would be a bit easier to review if all the old tests continued exhibiting the same behavior. An easy way to do this (I think) would be to continue to successfully typecheck a pattern-guarded RHS if it only has a single case.